framework.go

```diff
@@ -0,0 +1,171 @@
+//
+// A timeout can be provided via the given `ctx` value by using `context.WithTimeout()`
+func (f *Framework) ApplyCluster(ctx context.Context, cluster *application.Cluster) (*client.Client, error) {
+	clusterApplication, clusterCM, defaultAppsApplication, defaultAppsCM, err := cluster.Build()
+	if err != nil {
+		return nil, fmt.Errorf("failed to build cluster app: %v", err)
+	}
+
+	// Apply Cluster resources
+	if err := f.MC().Client.Create(ctx, clusterCM); err != nil {
+		return nil, fmt.Errorf("failed to apply cluster configmap: %v", err)
+	}
+	if err := f.MC().Client.Create(ctx, clusterApplication); err != nil {
+		return nil, fmt.Errorf("failed to apply cluster app CR: %v", err)
+	}
+	// Apply Default Apps resources
+	if err := f.MC().Client.Create(ctx, defaultAppsCM); err != nil {
+		return nil, fmt.Errorf("failed to apply default-apps configmap: %v", err)
+	}
+	if err := f.MC().Client.Create(ctx, defaultAppsApplication); err != nil {
+		return nil, fmt.Errorf("failed to apply default-apps app CR: %v", err)
+	}
+
+	return f.WaitForClusterReady(ctx, cluster.Name, cluster.Namespace)
+}
+
+// WaitForClusterReady watches for a Kubeconfig secret to be created on the MC and then waits until that cluster's api-server response successfully
+//
+// A timeout can be provided via the given `ctx` value by using `context.WithTimeout()`
+func (f *Framework) WaitForClusterReady(ctx context.Context, clusterName string, namespace string) (*client.Client, error) {
+	// Allow for context-based timeout handling
+	for {
+		select {
+		case <-ctx.Done():
+			// Timeout / deadline reached
+			return nil, ctx.Err()
+		default:
+			f.Log("Checking for valid Kubeconfig for cluster %s", clusterName)
+
+			var kubeconfigSecret corev1.Secret
+			err := f.MC().Get(ctx, types.NamespacedName{Name: fmt.Sprintf("%s-kubeconfig", clusterName), Namespace: namespace}, &kubeconfigSecret)
+			if cr.IgnoreNotFound(err) != nil {
+				return nil, err
+			} else if apierrors.IsNotFound(err) {
+				// Kubeconfig not yet available
+				f.Log(" - kubeconfig secret not yet available.\n")
+				time.Sleep(10 * time.Second)
+				continue
+			}
+
+			if len(kubeconfigSecret.Data["value"]) == 0 {
+				// Kubeconfig data not yet available
+				f.Log(" - kubeconfig secret not yet populated.\n")
+				time.Sleep(10 * time.Second)
+				continue
+			}
+
+			kubeconfig := string(kubeconfigSecret.Data["value"])
+			wcClient, err := client.NewFromRawKubeconfig(string(kubeconfig))
+			if err != nil {
+				return nil, err
+			}
+
+			if err := wcClient.CheckConnection(); err != nil {
+				// Cluster not yet ready
+				f.Log(" - connection to api-server not yet available.\n")
+				time.Sleep(10 * time.Second)
+				continue
+			}
+
+			f.Log(" - Got it!\n")
+
+			// Store client for later
+			f.wcClients[clusterName] = wcClient
+
+			return wcClient, nil
+		}
+	}
+}
+
+// DeleteCluster removes the Cluster app from the MC
```

Comment:
I think this would leave the configmaps as leftovers. I'm not sure about the default-apps app bundle.

Reply:
This is the same behaviour as the current tests. Do you think this should also remove the configmaps?

Reply:
I haven't updated the PR yet but my latest changes I've been working on do allow for creating new org and namespace per run so would clean these up too.